src/core_plugins/kibana/public/dashboard/panel/panel_header/panel_header_container.js

```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+
+import { PanelHeader } from './panel_header';
+import { PanelOptionsMenuContainer } from './panel_options_menu_container';
+import { PanelMaximizeIcon } from './panel_maximize_icon';
+import { PanelMinimizeIcon } from './panel_minimize_icon';
+import { DashboardViewMode } from '../../dashboard_view_mode';
+
+import {
+  maximizePanel,
+  minimizePanel,
+} from '../../actions';
+
+import {
+  getEmbeddable,
+  getEmbeddableTitle,
+  getMaximizedPanelId,
+  getFullScreenMode,
+  getViewMode
+} from '../../selectors';
+
+const mapStateToProps = ({ dashboard }, { panelId }) => {
+  const embeddable = getEmbeddable(dashboard, panelId);
+  return {
+    title: embeddable ? getEmbeddableTitle(dashboard, panelId) : '',
+    isExpanded: getMaximizedPanelId(dashboard) === panelId,
+    isViewOnlyMode: getFullScreenMode(dashboard) || getViewMode(dashboard) === DashboardViewMode.VIEW,
+  };
+};
+
+const mapDispatchToProps = (dispatch, { panelId }) => ({
+  onMaximizePanel: () => dispatch(maximizePanel(panelId)),
+  onMinimizePanel: () => dispatch(minimizePanel()),
+});
+
+const mergeProps = (stateProps, dispatchProps, ownProps) => {
+  const { isExpanded, isViewOnlyMode, title } = stateProps;
+  const { onMaximizePanel, onMinimizePanel } = dispatchProps;
+  const { panelId, embeddableHandler } = ownProps;
+  let actions;
+
+  if (isViewOnlyMode) {
+    actions = isExpanded ?
+      <PanelMinimizeIcon onMinimize={onMinimizePanel} /> :
+      <PanelMaximizeIcon onMaximize={onMaximizePanel} />;
+  } else {
+    actions = (
+      <PanelOptionsMenuContainer
+        panelId={panelId}
+        embeddableHandler={embeddableHandler}
+      />
+    );
+  }
+
+  return {
+    title,
+    actions,
+  };
+};
+
+export const PanelHeaderContainer = connect(
+  mapStateToProps,
+  mapDispatchToProps,
+  mergeProps,
+)(PanelHeader);
+
+PanelHeaderContainer.propTypes = {
+  panelId: PropTypes.string.isRequired,
+  /**
+   * @type {EmbeddableHandler}
```

Comment:
`PropTypes.shape` eh? Nicely done. I really need to get in the habit of not simply using `PropTypes.object`...

What's this `@type` doing? Some typescript stuff?

Reply:
not typescript, thats my attempt at jsdoc'ing this.  I can see the shape getting out of sync because it's repeated many places, but there is only one spot I define the jsdoc type EmbeddableHandler, where the base class is defined, so I wanted to indicate that's the object that will be passed in (without replicating every parameter we add to it on every shape).  I was thinking I'd mainly put the required params into the shape, but the jsdoc type will include all types.

I think all this will be obviated when we do introduce typescript officially.